### ISSUE TITLE
file exists in /dev/shm

### ISSUE THREAD

**Issue by botteon** [mhardcastle/ddf-pipeline#223]
Hi, 

my run crashes at the image_full_ampphase_di_m.NS step with this error:

```
 - 05:26:30 - ClassCasaImage            | [114.7/137.0 116.6/164.5 97.9Gb]   ----> Save image data as FITS file image_full_ampphase_di_m.NS.mask01.fits
 - 05:26:46 - ClassCasaImage            | [112.9/137.0 114.8/164.5 97.9Gb]   ----> Save image data as FITS file image_full_ampphase_di_m.NS.noise01.fits
 - 05:27:00 - ClassCasaImage            | [112.9/137.0 114.8/164.5 97.9Gb]   ----> Save image data as FITS file image_full_ampphase_di_m.NS.brutalRestored01.fits
 - 05:27:08 - ClassCasaImage            | [112.9/137.0 114.8/164.5 97.9Gb]   ----> Save image data as FITS file image_full_ampphase_di_m.NS.brutalModelConv01.fits
 - 05:27:11 - ClassModelMachineSSD      | [111.2/137.0 113.1/164.5 97.9Gb] Reference frequency already set to 144.627380 MHz
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]   Running minor cycle [MinorIter = 0/1000000, SearchMaxAbs = 1]
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]     Dirty image peak flux      =   0.022375 Jy [(min, max) = (-0.0224, 0.0122) Jy]
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]       RMS-based threshold      =   0.000138 Jy [rms = 4.6e-05 Jy; RMS factor 3.0]
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]       Sidelobe-based threshold =   0.000000 Jy [sidelobe  = 0.025 of peak; cycle factor 0.0]
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]       Peak-based threshold     =   0.000022 Jy [0.001 of peak]
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]       Absolute threshold       =   0.000000 Jy
 - 05:27:13 - ClassImageDeconvMachine   | [114.7/137.0 116.6/164.5 97.9Gb]     Stopping flux              =   0.000138 Jy [0.006 of peak ]
 - 05:27:15 - ClassIslandDistanceMachin | [114.7/137.0 116.6/164.5 97.9Gb] Searching Islands
 - 05:28:13 - ClassIslandDistanceMachin | [117.1/137.0 119.0/164.5 97.9Gb]   found 52793 islands
 - 05:28:15 - ClassImageDeconvMachine   | [114.9/137.0 116.8/164.5 97.9Gb]   selected 38565 islands [out of 52793] with peak flux > 0.000138 Jy
 - 05:28:15 - ClassIslandDistanceMachin | [114.9/137.0 116.8/164.5 97.9Gb]   calculating global islands cross-contamination
 - 05:28:15 - ClassIslandDistanceMachin | [114.9/137.0 133.5/164.5 97.9Gb]   extracting Island edges
    Calc. Dist. ............38565/38565 [================================================] 100%  - 59'37"
 - 06:29:54 - ClassIslandDistanceMachin | [143.1/158.3 150.1/165.4 97.9Gb]   grouping cross contaminating islands...
 - 06:30:09 - ClassIslandDistanceMachin | [148.6/158.3 155.6/165.4 97.9Gb]     have grouped 38565 --> 33037 islands
 - 06:30:09 - ClassIslandDistanceMachin | [148.6/158.3 155.6/165.4 97.9Gb]   Convexify islands
 - 06:30:21 - ClassIslandDistanceMachin | [148.7/158.3 155.7/165.4 97.9Gb]   Merge intersecting islands
 - 06:38:57 - ClassIslandDistanceMachin | [148.9/158.3 155.9/165.4 97.9Gb]   32953 islands remaining after merge
 - 06:38:58 - ClassIslandDistanceMachin | [148.8/158.3 155.9/165.4 97.9Gb]   calculating label image
 - 06:39:00 - ClassImageDeconvMachine   | [149.3/158.3 157.6/165.4 97.9Gb] Sorting islands by size
 - 06:39:05 - ClassImageDeconvMachine   | [115.6/158.3 118.8/165.4 97.9Gb]   Running minor cycle [MinorIter = 0/1000000, SearchMaxAbs = 1]
 - 06:39:36 - ClassImageDeconvMachine   | [119.0/158.3 122.3/165.4 97.9Gb]   selected 4662 islands larger than 10 pixels for initialisation
 - 06:39:36 - ClassInitSSDModel         | [119.0/158.3 122.3/165.4 97.9Gb] Initialise HMP machine
 - 06:39:36 - ClassPSFServer            | [119.0/158.3 122.3/165.4 97.9Gb] Using peak-normalised PSFs
 - 06:39:36 - CacheManager              | [119.0/158.3 122.3/165.4 97.9Gb] cache element ./big-mslist.txt.ddfcache/HMP_Init will be explicitly reset
 - 06:39:36 - ClassMultiScaleMachine    | [119.0/158.3 122.3/165.4 97.9Gb] using auto PSF box of size 405x405 in minor cycle subtraction
 - 06:39:37 - ClassMultiScaleMachine    | [110.3/158.3 113.6/165.4 97.9Gb] 48 scales and 7 scale functions in list
    Init HMP................... 109/109 [==================================================] 100%  - 0'14"
 - 06:39:53 - AsyncProcessPool          | [110.4/158.3 113.7/165.4 105.2Gb] Init HMP: 109 jobs complete, average single-core time 1.79s per job
 - 06:39:57 - ClassInitSSDModel         | [113.9/158.3 123.7/165.4 108.7Gb] Initialise islands (parallelised over islands)
    Init islands..............2596/4662 [===========================                       ]  55%  - 127'24"
    Init islands..............4662/4662 [==================================================] 100%  - 170'39"
 - 09:30:38 - AsyncProcessPool          | [113.9/158.3 123.7/165.4 108.8Gb] Init islands: 4662 jobs complete, average single-core time 30.58s per job
 - 09:30:40 - ClassImageDeconvMachine   | [107.0/158.3 116.7/165.4 108.8Gb] Evolving 50 generations of 50 sourcekin
 - 09:30:40 - ClassImageDeconvMachine   | [107.0/158.3 116.7/165.4 108.8Gb] Deconvolve small islands (<=10000 pixels) (parallelised over island)
Traceback (most recent call last):
  File "/opt/lofar/DDFacet/bin/DDF.py", line 387, in <module>
    main(OP, messages)
  File "/opt/lofar/DDFacet/bin/DDF.py", line 235, in main
    Imager.main()
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Imager/ClassDeconvMachine.py", line 1097, in main
    repMinor, continue_deconv, update_model = self.DeconvMachine.Deconvolve()
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Imager/SSD/ClassImageDeconvMachineSSD.py", line 409, in Deconvolve
    self.DeconvListIsland(ListSmallIslands,ParallelMode="OverIslands",ListInitIslands=ListInitSmallIslands)
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Imager/SSD/ClassImageDeconvMachineSSD.py", line 489, in DeconvListIsland
    island_dict["Island"] = np.array(ThisPixList)
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Array/shared_dict.py", line 255, in __setitem__
    value = NpShared.ToShared(_to_shm(path+'a'), value)
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Array/NpShared.py", line 81, in ToShared
    a = CreateShared(Name, A.shape, A.dtype)
  File "/opt/lofar/DDFacet/lib/python2.7/site-packages/DDFacet/Array/NpShared.py", line 76, in CreateShared
    a = SharedArray.create(Name, shape, dtype=dtype)
OSError: [Errno 17] File exists: 'file:///dev/shm/ddf.26722/DeconvListIslands/int:29681:d/str:Island:a'
```

I removed all the files in /dev/shm, but it crashed with the same error.

Notes on the run: 
-deep field (223 ms files)
-imsize 21500 (but the node that I'm using already made images so big with DDF)
-this https://github.com/mhardcastle/ddf-pipeline/issues/161 occurred twice and was solved by clearing the cache

What is going on?
Thanks

**Comment by mhardcastle** [comment]
This 'file exists' error is quite obscure but seems to correlate with having lots of MSs. The suggested workround in recent DDF versions is
```
 - 08:07:53 - DDFacet            | WARNING: sysctl vm.max_map_count = 65530. 
            This may be too little for large DDFacet and killMS jobs. If you get strange "file exists" 
            errors on /dev/shm, them try to bribe, beg or threaten your friendly local sysadmin into 
            setting vm.max_map_count=1000000 in /etc/sysctl.conf.
```
Can you do `sysctl -a | grep max_map` and see what your local setting is? And then see if you can get a sysadmin to change it?

**Comment by botteon** [comment]
Ok, thanks. I will contact the system admins (`sysctl` is not installed in Leiden and `/etc/sysctl.conf` doesn't show any custom setting).

I'll close the issue when I pass the step. Should I start the new attempt with `--control-clearcache=True`?

**Comment by mhardcastle** [comment]
Might be in /usr/sbin or /sbin, which may not be on your path? But anyway sysadmins will be needed to make the change.

You should not need to clear the ddfcache files on restart.

**Comment by botteon** [comment]
Indeed, it was not on the path.

```
>/usr/sbin/sysctl -a | grep max_map
vm.max_map_count = 65530
```

**Comment by botteon** [comment]
Changing `vm.max_map_count` to 1000000 was not enough, and the node started swapping.
@twshimwell suggested me to append `--Misc-ConserveMemory=1` to the DDF call in `pipeline.py` (following https://github.com/mhardcastle/ddf-pipeline/issues/150).
I had to clear the memory, that's why it took 2 days, but now the final image is done, and it was worth the wait :)

**Comment by mhardcastle** [comment]
To  be honest I am not sure what the Misc-ConserveMemory option does, but we could make this an option in the pipeline (under 'control') if it might be useful to others in future.

**Comment by cyriltasse** [comment]
`--Misc-ConserveMemory=1` loads the weights in series if I remember correctly...

> I had to clear the memory, that's why it took 2 days, but now the final image is done, and it was worth the wait :)

Nice to hear! :+1:

**Comment by mariasde** [comment]
Hi,

I just wanted to say that I've also been working on a deep field (the North Ecliptic Pole) and I ran into a similar issue as Andrea. In my case I didn't get an error message, but the pipeline would just stall in the stage
- ClassImageDeconvMachine   | [35.6/88.0 53.3/104.1 162.3Gb] Deconvolve small islands (<=10000 pixels) (parallelised over island)
and in particular
Init islands..............5152/5156

In the end I did as Andrea and set vm.max_map_count to 1000000 and appending --Misc-ConserveMemory=1 to the DDF call in pipeline.py and that worked. I thought you might want to know, if you're thinking of making Misc-ConserveMemory an option in the pipeline.

Cheers,
María
